### PR TITLE
Add a check in the task runner to wait for dependents to finish

### PR DESCRIPTION
**Pull request by i-like-robots**
## What

This PR adds logic to improve the usage of the `--concurrency` option with a new `--preserve-order` flag.

## Why

Running concurrent tasks is great, but not always. This is because tasks can be kicked off for packages which may depend on another package that has not yet finished running... e.g.:

- There are two modules: A and B
- A depends on B
- Both modules have a build step
- Running a build task with a concurrency of 2 would immediately kick off both builds
- The task will fail because A cannot find B

## How

This PR adds an additional queue to the semaphore concurrency pattern which ensures tasks can wait until any dependent packages finish running. There are a number of changes to aid this:

1. I have refactored each action to return the package info along with the function to run the task. This was done in order to access the package name and dependencies.
2. I have added an evented queue to the "run-parallel.js" module. This will maintain a list of packages with currently queued and still running tasks. The evented queue can be asked for a promise which resolves when the queue no longer contains any dependent packages.
3. I have added an `allDependencies` property to the package wrapper class as this was duplicated in several places
4. I have refactored the tests to use a shared helper in order to create mock package wrappers

I tested by running various tasks within the the x-dash repository.

## Show

Usage:

```sh
# long
athloi run build --concurrency 5 --preserve-order

#short
athloi run build -C 5 -P
```

### CODE REVIEW

File: src/run-parallel.js [hunk 1]
```diff
@@ -1,14 +1,35 @@
...
+	const packagesRunning = new Set();
+
 	return Promise.all(
-		tasks.map((task) => {
-			return sema.acquire().then(task).then(() => sema.release());
+		tasks.map(({ pkg, apply }) => {
+			return semaphore
+				.acquire()
+				.then(() => {
+					return waitUntil(() => {
```
Comment (i-like-robots) [inline src/run-parallel.js]: Perhaps this may be neater with an event emitter rather than a poll?

File: readme.md [hunk 2]
```diff
@@ -93,18 +93,25 @@ athloi publish -- --access=public
 
 ### concurrency
 
-A global concurrency option which can be used to execute multiple tasks in parallel. By default only one task will run at a time.
+A global option which will execute up to the given number of tasks concurrently. By default one task will be run at a time.
```
Comment (pixelandpage) [inline readme.md]: Should there be an upper-limit to the number of tasks we can run concurrently? 🤔

File: test/src/evented-queue.spec.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		instance = new Subject();
+	});
+
+	describe('#add', () => {
+		it('adds the given item to the queue', () => {
+			instance.add('foo');
+			expect(instance.queue.size).toEqual(1);
+		});
+
+		it('emits an event when items are added to the queue', (done) => {
```
Comment (pixelandpage) [inline test/src/evented-queue.spec.js]: should this test have an 'expect' declaration?

File: test/src/evented-queue.spec.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	describe('#delete', () => {
+		it('removes the given item to the queue', () => {
+			instance.add('foo');
+			expect(instance.queue.size).toEqual(1);
+
+			instance.delete('foo');
+			expect(instance.queue.size).toEqual(0);
+		});
+
+		it('emits an event when items are removed from the queue', (done) => {
```
Comment (pixelandpage) [inline test/src/evented-queue.spec.js]: should this test have an 'expect' declaration?

File: test/src/evented-queue.spec.js (hunk 3 above)
Comment (i-like-robots) [reply test/src/evented-queue.spec.js]: Good question! But no it shouldn't. The `done()` callback in this case is the success criteria, failure would be not calling it.

**Pull request merged**